kobo/client/__init__.py

```diff
@@ -300,6 +285,64 @@ def get_server_principal(service=None, realm=None):
 
         self._hub.auth.login_krbv(req_enc)
 
+    def _login_gssapi(self):
+        """Login using kerberos credentials (uses gssapi)."""
+        # read default values from settings
+        principal = self._conf.get("KRB_PRINCIPAL")
+        keytab = self._conf.get("KRB_KEYTAB")
+        service = self._conf.get("KRB_SERVICE")
+        realm = self._conf.get("KRB_REALM")
+        ccache = self._conf.get("KRB_CCACHE")
+        proxyuser = self._conf.get("KRB_PROXYUSER")
+
+        import enum
+        import gssapi
+
+        if principal is not None:
+            if keytab is None:
+                raise ImproperlyConfigured("Cannot specify a principal without a keytab")
+            name = gssapi.Name(principal, gssapi.NameType.kerberos_principal)
+            store = {'client_keytab': keytab}
+            if ccache is not None:
+                store['ccache'] = 'FILE:' + ccache
+            
+            client_credentials = gssapi.Credentials(name=name, store=store, usage='initiate',
+                                                    mechs=gssapi.MechType.kerberos)
+        else:
+            if ccache is not None:
+                store = {'ccache': 'FILE' + ccache}
+                client_credentials = gssapi.Credentials.acquire(usage='initiate', store=store,
+                                                                mechs=gssapi.MechType.kerberos)
+            else:
+                client_credentials = gssapi.Credentials.acquire(usage='initiate',
+                                                                mechs=gssapi.MechType.kerberos)
+
+        server_name = self.get_server_principal(service=service, realm=realm)
+        flags = 0x00000004 | 0x00000001
+        client_context = gssapi.SecurityContext(name=server_name, creds=client_credentials,
+                                                flags=flags, usage='initiate')
+        try:
+            output_token = client_context.step()
+            self._hub.auth.login_gssapi(output_token)
+        except gssapi.exceptions.GSSError as e:
+            if getattr(e, "min_code", None) == 2529638919:
+                e.message += ". Make sure you correctly set KRB_REALM (current value: %s)." % realm
+                e.args = (e.min_code, e.message)
+            raise e
+
+    def get_server_principal(service=None, realm=None):
```

Comment:
I decided to give this patch a try and I found that it doesn't work.  I think you forgot to put either a `@staticmethod` or a `self` argument in this method, therefore it always fails with:

```
Failed to create new session: get_server_principal() got multiple values for argument 'service'
```

...because the "service" argument is used as self, but the caller also provides a "service".

Can you please fix that?  I'll test it again afterward.

Also, please look into adding relevant tests, a trivial bug such as this shouldn't have to be caught by a human reviewer.

Reply:
I have fixed the missing parameter. My bad.

Unfortunately, no test suite currently exists for the log-in methods. So creating tests for gssapi would probably require significant effort. And I'm not sure if unit tests are well suited for this kind of code, since it's mostly just API calls, most of which would have to be mocked.